DM-48013: Add contribution guide and pull request template

JeremyMcCormick opened the pull request:
Added:

- Contribution guide at the standard location of `CONTRIBUTING.md`
- New pull request template with:
  - Checkbox for incrementing the schema version number, if applicable
  - Checkbox for referring to schema-specific information that may need to be consulted
- APDB documentation under `docs/APDB.md`, which includes versioning recommendations and other information relevant to contributing
 
It is planned to add more schema-specific documentation under `docs` for all major schemas, but probably not on this PR. Documentation for APDB is included here as a working case for which information to put in the main guide and what to put under the schema-specific pages.

----

The new contribution guide can be previewed [here](https://github.com/lsst/sdm_schemas/blob/tickets/DM-48013-jeremym/CONTRIBUTING.md).

----

An APDB-specific schema versioning guide by @andy-slac which was used for the content in `docs/APDB.md` can be found [here](https://rubinobs.atlassian.net/wiki/external/OGUyOGM1OTY0MTRkNGFmNThkNWM4N2U1NGFmODJkNWY).

kfindeisen commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+| Change | Increment | Notes |
+|--------|-----------|-------|
+| Adding a table | `MAJOR` | Clients will want to access the new table, so this is backward incompatible.
+| Adding a column | `MAJOR` | Clients will want to access the new column, so this is incompatible with the existing schema.|
+| Removing a column | `MINOR` | Clients should not care about a missing column, so this should be backward compatible.|
```
These seem a bit broader and harder to visualize than the original text. Perhaps re-emphasize that this is talking about _new_ clients that already have had references to the column added/removed?

Actually, I think that may be the key issue. The original docs were specific to the APDB, which is accessed almost exclusively through the `dax_apdb` API. Does the same go for other tables, or is there more room for e.g. references to a removed column?

JeremyMcCormick replied:
> These seem a bit broader and harder to visualize than the original text. Perhaps re-emphasize that this is talking about _new_ clients that already have had references to the column added/removed?
> 
> Actually, I think that may be the key issue. The original docs were specific to the APDB, which is accessed almost exclusively through the `dax_apdb` API. Does the same go for other tables, or is there more room for e.g. references to a removed column?

I can clarify this text. You're completely right that a client would need to be updated if a column was dropped from the schema and it was being accessed explicitly.

JeremyMcCormick replied:
The other option is that the guidelines/rules for the versioning could go under the APDB-specific documentation, though I would like to provide some general recommendations for any versioned schema, if possible.

JeremyMcCormick replied:
I moved these guidelines to `docs/APDB.md` as they are APDB-specific.

kfindeisen replied:
> Some of these I'm not even sure I agree with - wouldn't removing a column be a backward-incompatible change necessitating a MAJOR version bump since old clients could be attempting to access it?

In this context @andy-slac is specifically defining "backward compatible" as "new client with old DB". I agree it's confusing, but it's at least internally consistent.

JeremyMcCormick replied:
> > Some of these I'm not even sure I agree with - wouldn't removing a column be a backward-incompatible change necessitating a MAJOR version bump since old clients could be attempting to access it?
> 
> In this context @andy-slac is specifically defining "backward compatible" as "new client with old DB" (that definition seems to have stayed in the generic docs and not the APDB-specific ones?). I agree it's confusing, but it's at least internally consistent.

Okay, thank you for clarifying. This is usually defined the other way around, but I moved this into the APDB-specific docs so that it was clear this is not a general rule for versioning all schemas.

JeremyMcCormick replied:
> that definition seems to have stayed in the generic docs and not the APDB-specific ones?

Good catch - I'm going to move this under the APDB docs as well. Each schema which uses versioning is probably going to handle things differently.

kfindeisen commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+Specific Schema Documentation
+-----------------------------
+
+- [Alert Production Database (APDB)](docs/APDB.md)
```
Broken link.

JeremyMcCormick replied:
Thanks for looking. This is a draft and work in progress. I will add it soon. :)

JeremyMcCormick commented:
@kfindeisen I'm going to add a short section mentioning that APDB changes must be propagated to api_package. If you have specific language that you would like to use, please let me know.

In terms of PR mechanics, does a merged PR on `apdb.yaml` trigger the updates to `api_package` which could come later? Or is this supposed to be more-or-less simultaneous?

JeremyMcCormick commented:
Andy's APDB-specific schema versioning page, for reference:

https://rubinobs.atlassian.net/wiki/external/OGUyOGM1OTY0MTRkNGFmNThkNWM4N2U1NGFmODJkNWY

JeremyMcCormick commented:
The new docs can be viewed here:

https://github.com/lsst/sdm_schemas/blob/tickets/DM-48013-jeremym/CONTRIBUTING.md

kfindeisen commented:
> @kfindeisen I added a short section in the APDB docs mentioning that changes must be propagated to `alert_packet`. If you have specific language that you would like to include there, please let me know.

I have only a very rough knowledge of `alert_packet`; please direct those questions to @ebellm.

JeremyMcCormick commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,48 @@
+Schema Versioning
+-----------------
+
+A pull request may necessitate an update to the schema's version.
```
I reworked the introductory paragraph of the versions section to incorporate these suggestions.

kfindeisen reviewed (changes requested):
I think this needs more coordination (with @andy-slac?) over what changes actually cause breakages. In particular, the primary table of examples assumes that DBs always use the latest schema, which in my experience is not true (migration support is often a low priority).

I also found quite a few typos and broken links, but I don't expect any controversy there.

andy-slac commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,97 @@
+| Remove column | MAJOR | Removing a column breaks queries and code expecting that column, thus not backward compatible. |
+| Change column type| MAJOR | Changing a column’s data type can break clients expecting a different type, forcing a major version increment. |
+| Add index | PATCH | Indexes improve performance without changing the schema’s interface, making this a safe, backward-compatible improvement. |
+| Add constraint | MAJOR | New constraints can invalidate previously acceptable data, breaking clients that rely on the old lenient rules. |
+| Change constraint | MAJOR | Altered constraints can break clients relying on the previous constraint behavior, making this a non-backward-compatible change. |
+| Remove constraint | MINOR | Removing a constraint generally relaxes restrictions. It’s unlikely to break clients, but it changes behavior enough to warrant more than a patch. |
```
Constraint changes are not at all clear case. These only affect database itself, client code usually does not care about constraint definitions. I guess you may be thinking in the terms of semver which is not quite applicable to these cases. It may be better to say that it should be decided on a case-by-case basis.

JeremyMcCormick replied:
Removed this section entirely.

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,47 @@
+The [apdb](./python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB).
+This database is updated by the [Alert Production Pipeline](https://github.com/lsst/ap_pipe) and used for nightly [ap_verify](https://github.com/lsst/ap_verify) runs within continuous integration.
+Previous processing runs may differ from the current schema.
+The user-queryable Prompt Products Database (PPDB) is expected to have a very similar schema to the APDB.
```
I don't think we need to say anything about PPDB here, that one is still more or less unknown w.r.t. schema management.

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,47 @@
+Alert Production Database (APDB) Schema
+=======================================
+
+The [apdb](./python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB).
+This database is updated by the [Alert Production Pipeline](https://github.com/lsst/ap_pipe) and used for nightly [ap_verify](https://github.com/lsst/ap_verify) runs within continuous integration.
+Previous processing runs may differ from the current schema.
```
I don't think you need to explain what APDB is and how it is populated in this document.

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,47 @@
+The deployed database schema is a product of the [schema in sdm_schemas](yml/apdb.yaml) and processing by tools in the [dax_apdb](https://github.com/lsst/dax_apdb) repository.
+Compatibility of the client executable with the actual database schema is determined by the version number in `apdb.yaml` and that which is stored in the database metadata.
+
+Backward compatibility in the context of the APDB implies that a client using an interface derived from a schema with a higher `MINOR` version number will be able to read and write into a database with a lower `MINOR` version.
```
Hmm, I don't think I put it like that on the Confluence page. In APDB there is no "interface derived from a schema". There is just schema used by the client to run the queries and an actual schema in the database. It may be better to copy that page more or less verbatim, I think it provides clearer explanation of what compatibility means for APDB?

kfindeisen replied:
I'm guessing that was a reference to `Apdb`, since I mentioned earlier that we never do direct queries.

JeremyMcCormick replied:
I attempted to reword this section so it was clearer and matched your text better.

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,47 @@
+Reviewers
+---------
+
+- [Andy Salnikov](https://github.com/andy-slac)
```
I should not be the only reviewer for APDB schema changes, but I can be added to review a corresponding version change.

JeremyMcCormick replied:
This is more like who to select as a default if it is otherwise unknown.

andy-slac replied:
This is still not right, I would know even less than PR author whom to call. Maybe add @isullivan instead or whomever Ian recommends.

andy-slac reviewed (changes requested):
I think there is a lot of confusion about the role of the version numbers. I agree with Krzysztof that we should stop calling it SemVer. I think people used to think about version in terms of APIs, where "server" side can be updated before the "client" side (e.g. dynamic library is updated, which can cause rebuilding of the client apps). In our case the roles are reversed - we update sdm_schemas which is "immediately" picked up by the database client code, and only later we can decide to upgrade the actual database schema. 

One more thing that I believe is implied in your description of schema compatibility is that the database schema version should be tracked somewhere (in the database itself), but there is no explicit mentioning of that anywhere.

kfindeisen commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,97 @@
+If a pull request shows the message "This branch is out-of-date with the base branch", then the "Update with rebase" button should be selected and pressed on the right to update it.
+The "Update branch" selection should not be used, as it will perform a merge, which is undesirable behavior that will cause the rebase check to fail.
+
+Rebasing can also be accomplished using a command like `git rebase -i main` locally after updating `main` and then force pushing the rebased branch using `git push -f`.
```
Personally, I do think `-i` is good practice. I've seen Git get confused with nontrivial edit histories, particularly branches that were once based off of other branches.

But that's not relevant to conflict resolution -- `-i` only means that you see (and edit) the list of commits Git plans to apply beforehand.

kfindeisen replied:
Personally, I do think `-i` is good practice. I've seen Git get confused with nontrivial edit histories, particularly branches that were once based off of other branches.

But that's not relevant to conflict resolution -- `-i` only means that you see (and edit) the list of commits Git plans to apply beforehand.

andy-slac replied:
Well, in your example the conflict would be noticeable during rebase. Suppose that both changes from 1.2.3 to 2.0.0, then there will be no rebase conflict. But in this case the second branch would need to update it to 3.0.0 after rebase.

JeremyMcCormick replied:
I just deleted this section entirely.

Rebasing is covered by DM documentation, and version conflicts from rebasing can be worked out on a case-by-case basis. (I can add a sentence or two about this in the section on versioning though, if you think it is worth noting.)

kfindeisen commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,50 @@
+Schema Versioning
+-----------------
+
+Individual schemas may have their own internal [version](https://felis.lsst.io/user-guide/model.html#schema-version) for tracking changes, which is defined at the top level of the file, as in [apdb.yaml](yml/apdb.yaml).
```
But what's important in this context is where a _developer_ can find the version number in order to check and/or change it. (Getting a strange sense of deja vu...)

JeremyMcCormick commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,97 @@
+
+Below are some general guidelines that can be used for incrementing the schema version based on what changes are being made:
+
+| Operation | Increment | Notes |
+|--------|-----------|-------|
+| Add table | MAJOR | Even if backward compatible, new tables often signal major functionality changes, usually requiring clients to update their code or queries. |
+| Remove table | MAJOR | Removing a table may break existing queries and client interfaces, making it incompatible with previous versions. |
+| Add column | MAJOR | Adding a column may require clients to adapt to the new data structure (e.g., parsing new fields), treated as a major change for consistency. |
+| Remove column | MAJOR | Removing a column breaks queries and code expecting that column, thus not backward compatible. |
+| Change column type| MAJOR | Changing a column’s data type can break clients expecting a different type, forcing a major version increment. |
+| Add index | PATCH | Indexes improve performance without changing the schema’s interface, making this a safe, backward-compatible improvement. |
+| Add constraint | MAJOR | New constraints can invalidate previously acceptable data, breaking clients that rely on the old lenient rules. |
+| Change constraint | MAJOR | Altered constraints can break clients relying on the previous constraint behavior, making this a non-backward-compatible change. |
+| Remove constraint | MINOR | Removing a constraint generally relaxes restrictions. It’s unlikely to break clients, but it changes behavior enough to warrant more than a patch. |
+| Change metadata | PATCH | Adjusting metadata that does not affect the schema’s functional behavior or data shape is a minor adjustment that remains backward compatible. |
```
More typical ways of thinking about compatibility between a database schema and its clients would be:

- The database was updated to a new schema, but a downstream client was not.
- SQL queries which have not been updated are being run on the updated schema.

What you're describing sounds like a technicality of APDB, which is why its sub-page has different versioning guidelines, if an updated client being run against a prior version of the database is the biggest concern.

That said, I'm willing to remove this section entirely if there is little to no commonality in how different schemas in the repo are or will be versioned. That information could be left up to schema-specific documentation.

JeremyMcCormick replied:
I deleted it. 

The APDB-specific guidelines should be consulted for specific versioning rules when changing `apdb.yaml`.

JeremyMcCormick commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,97 @@
+(This is distinct from tags or versions of sdm_schemas itself, and not all schemas may be using these internal versions.)
+This version may need to be updated when making changes.
+
+Usage of semantic versioning is recommended, which defines a version like `MAJOR.MINOR.PATCH`, such as `1.2.3`.
```
I could use a suggestion on how to describe the versioning, if you think this is misleading or incorrect. I'm not sure what to put, otherwise.

JeremyMcCormick commented:
@andy-slac 

> I agree with Krzysztof that we should stop calling it SemVer.

Based on this feedback, I reworded this section and removed the part about using using semantic versioning. Please have a look at the updated text.

> we update sdm_schemas which is "immediately" picked up by the database client code, and only later we can decide to upgrade the actual database schema

After thinking about this, I'm in line with this and I have updated the main text accordingly.

> the database schema version should be tracked somewhere (in the database itself), but there is no explicit mentioning of that anywhere

I plan to add this capability - see [DM-47340](https://rubinobs.atlassian.net/issues/DM-47340).

JeremyMcCormick commented:
@kfindeisen

> I think this needs more coordination (with @andy-slac?) over what changes actually cause breakages. In particular, the primary table of examples assumes that DBs always use the latest schema, which in my experience is not true (migration support is often a low priority).
> 

Thank you for the thorough review.

> what changes actually cause breakages

After discussing, I have realized that using an updated client on an older schema version is a more typical case than the reverse, so I will rewrite the main versioning "operations" table accordingly.

> I also found quite a few typos and broken links, but I don't expect any controversy there.

The ones you mentioned should all be fixed now. Please let me know if you find additional mistakes.

JeremyMcCormick commented:
> In particular, the primary table of examples assumes that DBs always use the latest schema, which in my experience is not true (migration support is often a low priority).

I don't think we can assume this as a general use case, as some other schema owners are considering it from the other direction. 

I have made the main contributing guide a lot more generic, assuming that the schema subpages will outline their own versioning guidelines. I have also removed the table from `CONTRIBUITNG` which was providing suggestions for specific operations like removing or adding a column.

ebellm commented on docs/APDB.md:
```diff
@@ -0,0 +1,49 @@
+Client Compatibility
+--------------------
+
+Changes to the APDB schema must be propagated to the API defined in the [api_packet](https://github.com/lsst/alert_packet) repository, following the instructions under [Adding a new schema](https://github.com/lsst/alert_packet/tree/main?tab=readme-ov-file#adding-a-new-schema).
```
```suggestion
Changes to the APDB schema must be propagated to the Avro alert schemas defined in the [alert_packet](https://github.com/lsst/alert_packet) repository, following the instructions under [Adding a new schema](https://github.com/lsst/alert_packet/tree/main?tab=readme-ov-file#adding-a-new-schema).
```

kfindeisen commented on docs/APDB.md:
```diff
@@ -0,0 +1,49 @@
+The APDB database schema is upgraded using a special migration tool defined in the [dax_apdb_migrate](https://github.com/lsst-dm/dax_apdb_migrate) repository.
+For every version change to `apdb.yaml`, a new migration script needs to be added to that repository.
+The details of the migration process are described in the [package documentation](https://pipelines.lsst.io/modules/lsst.dax.apdb/index.html).
+Migration scripts should normally be implemented on the same Jira ticket as the corresponding schema change.
```
@andy-slac, should this mention the distinction between SQL and Cassandra (or the fact that the latter is currently manual rather than scripted)?

JeremyMcCormick replied:
> the distinction between SQL and Cassandra

SQL = Postgres?

andy-slac replied:
I do not want to mention manual part here, it's mostly irrelevant, but mentioning Cassandra is probably worth it. @JeremyMcCormick, could you add a sentence here:
```
APDB has a Cassandra-based implementation in addition to a SQL-based, and Cassandra may impose additional restrictions on possible schema changes.
```

kfindeisen commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+Specific Schema Documentation
+-----------------------------
+
+- [Alert Production Database (APDB)](docs/APDB.md)
```
I meant to ask this last time but forgot: please add `APDB.md` on or before this commit. Otherwise, the link is temporarily broken.

kfindeisen commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,6 @@
+
+When making changes to YAML files in the [schemas](../python/lsst/sdm_schemas/schemas) directory:
+
+- [ ] Incremented the schema version number, following the guidelines in the [contribution guide](../CONTRIBUTING.md) (if applicable)
```
I think the following would be easier to understand:
```suggestion
- [ ] If applicable, incremented the schema version number, following the guidelines in the [contribution guide](../CONTRIBUTING.md)
```

kfindeisen reviewed (approved):
Looks much better, thanks! Just minor comments.

JeremyMcCormick commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+Schema Versioning
+-----------------
+
+Individual schemas may have their own internal [version](https://felis.lsst.io/user-guide/model.html#schema-version) for tracking changes, which is defined at the top of the file, as in [apdb.yaml](yml/apdb.yaml).
```
Fixed

andy-slac commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+Schema Versioning
+-----------------
+
+Individual schemas may have their own internal [version](https://felis.lsst.io/user-guide/model.html#schema-version) for tracking changes, which is defined at the top of the file, as in [apdb.yaml](./python/lsst/sdm_schemas/schemas/yml/apdb.yaml).
```
Link is incorrect.

```suggestion
Individual schemas may have their own internal [version](https://felis.lsst.io/user-guide/model.html#schema-version) for tracking changes, which is defined at the top of the file, as in [apdb.yaml](./python/lsst/sdm_schemas/schemas/apdb.yaml).
```

andy-slac commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- [Compare Schemas for Changes](.github/workflows/compare.yaml) - Schemas that are in the [deployed list](./python/lsst/sdm_schemas/schemas/deployed-schemas.txt) will fail this check.
+- [Build the Schema Browser website](.github/workflows/docs.yaml)
+- [Check that `main` is not merged into the branch](.github/workflows/rebase_checker.yaml)
+- [Test that the schemas can be used to create databases](.github/workflows/test_databases.yml)
```
```suggestion
- [Test that the schemas can be used to create databases](.github/workflows/test_databases.yaml)
```

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,49 @@
+Alert Production Database (APDB) Schema
+=======================================
+
+The [apdb](../python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB),
+which contains the results of processing from the [Alert Production Pipeline](https://github.com/lsst/ap_pipe).
```
```suggestion
The [apdb](../python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB), which contains the results of processing from the [Alert Production Pipeline](https://github.com/lsst/ap_pipe).
```

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,49 @@
+-----------------
+
+The versioning system of the APDB is outlined in [DMTN-269](https://dmtn-269.lsst.io/).
+The deployed database schema is a product of the [schema in sdm_schemas](../python/lsst/sdm_schemas/schemas/apdb.yaml) and processing by tools in the [dax_apdb](https://github.com/lsst/dax_apdb) repository.
```
```suggestion
The deployed database schema is a product of the [schema in sdm_schemas](../python/lsst/sdm_schemas/schemas/apdb.yaml) with additional processing by APDB client library in the [dax_apdb](https://github.com/lsst/dax_apdb) repository.
```

andy-slac commented on docs/APDB.md:
```diff
@@ -0,0 +1,49 @@
+
+The APDB database schema is upgraded using a special migration tool defined in the [dax_apdb_migrate](https://github.com/lsst-dm/dax_apdb_migrate) repository.
+For every version change to `apdb.yaml`, a new migration script needs to be added to that repository.
+The details of the migration process are described in the [package documentation](https://pipelines.lsst.io/modules/lsst.dax.apdb/index.html).
```
The link is incorrect. We do not have published docs for migration script yet, but better link should be:
```suggestion
The details of the migration process are described in the [package documentation](https://github.com/lsst-dm/dax_apdb_migrate/blob/main/doc/lsst.dax.apdb_migrate/index.rst).
```

andy-slac reviewed (approved):
Looks good, few suggestions.

gpdf reviewed (approved):
Looks good now.  Don't wait for schema-specific documentation for ConsDB, but let's make sure there's a ticket for adding that.
